### ISSUE TITLE
Shell autocompletion for arguments

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#302]
Would be nice to have that kind of thing for zsh and bash. Not sure how this can be done or f this can be done on our side.

**Comment by sils1297** [comment]
Maybe https://argcomplete.readthedocs.org/en/latest/ holds useful info.

**Comment by sharky93** [comment]
Basic argument auto completion can be triggered by registering a complete function such as `_longopt` with the `complete` built-in. It searches the `--help` output for completions.

For future use, we plan to add auto completion for bears too, that is still achievable by adding one or two more functions in the same fashion.
But there will have to be different solutions for `bash` and `zsh`. Instead of this, should we go by asking users to install a Python package such as `argcomplete`, which claims to have support for bash and zsh both.

**Comment by sils1297** [comment]
So for now you go with installing __longopt? Did I understand correctly?
Am 20.02.2015 15:47 schrieb "sharky93" <notifications@github.com>:

**Comment by sharky93** [comment]
`argcomplete`
Pros
1. claims support for bash, zsh
2. easy to use api

Cons
1. a local user install is tricky, does not work perfectly for me as mentioned in my comment above
2. further, it might just have support where in the `parse_args` module is called from directly within the script we are hooking the auto complete feature to, in our case we parse from an instance of another class, relevant question @ https://github.com/kislyuk/argcomplete/issues/96
3. need to install a full extra module

`a native bash solution`
Pros
1. Full freedom to modify as per our little/heavy requirements
2. No need to ask the user to install another module

Cons
1. Need to modify/create `~/bash_completion` to add the function which handles the auto complete feature. Is this too much to ask?

Hacked up a simple bash solution and it works well. Haven't figured out handling `zsh` but should be rather straightforward.

**Comment by sils1297** [comment]
With the native bash solution we're kind of reinventing the wheel, aren't we? I have nothing against installing some file in some _system_ directory but I wouldn't want to add something to a file lying in the _user_ directory.

Since we want to install it when running setup.py which is usually called with sudo anyway I don't see why we need a local user install.

Can you show your native bash solution so we can see how hacky it is? Is it possible to install the autocompletion system wide without changing user specific things? I think we usually want that.

**Comment by sharky93** [comment]
I'm all for not reinventing the wheel but in this case I feel a new wheel is giving us more control with very less additions since ours is a relatively simple use case.

Sure we can do a system wide install, by creating `/etc/bash_completion.d/coala`. I usually prefer local changes so that if things screw up, they don't affect all people but I don't have any other strong point to go with it.

Here's the native bash solution - https://gist.github.com/sharky93/8a5a96e4f6728fbc8322

**Comment by sils1297** [comment]
Well if the user installs  coala he wants the installation systemwide so he'll want the autocompletion systemwide and the he want nothing screwed up ;)

**Comment by sils1297** [comment]
If we can do that with that kind of script and without additional deps, it's great. We'll need to tweak the script a bit.

**Comment by sharky93** [comment]
Should we have this bash script as part of the repo as a file which we copy at the required destination OR have the contents inside the setup script and the create a new file?

**Comment by sils1297** [comment]
Having files in files as strings or so is usually not the cleanest solution. Did you see my comments at the script itself?

**Comment by AbdealiJK** [comment]
There seems to be a nice library to do this : https://pypi.python.org/pypi/argcomplete/0.8.4

**Comment by sharky93** [comment]
So, I experimented a bit on why `argcomplete` was not working out of the box, turns out translations of the help texts is causing it to crash.
Interestingly, on removing translations, a standalone version of the `default_arg_parser` works but within coala it still doesn't work implying that there is something more too that is causing an issue.

**Comment by sils1297** [comment]
@sharky93 any progress?

**Comment by sils1297** [comment]
Accepting patches.

**Comment by Uran198** [comment]
I think you can make `argcomplete` pluggable. All it requires is calling `argcomplete.autocomplete(parser)` before executing `args = parser.parse_args()`.
So, to make this optional and pluggable following lines can be added to `coalib/coala.py`:
```python
    try:
        import argcomplete
        argcomplete.autocomplete(arg_parser)
    except ImportError:
        pass
```
Though this will autocomplete options, to make it autocomplete bears will require more effort, but hopefully not a lot.

**Comment by Makman2** [comment]
Don't remember we closed this issue. Do we want to reopen, as this enhances user experience?
Especially if it works that easy like @Uran198 writes, we definitely should do that^^

**Comment by sils1297** [comment]
It's a won'tfix, reopen if somebody is doing that or we decide that we definitely want that and it's doable with an agreeable amount of work.

**Comment by sils1297** [comment]
and @Uran198 doesn't want to do it

**Comment by sils1297** [comment]
(at least not right now)

**Comment by Makman2** [comment]
6 lines of code is not an "agreeable amount of work"? :O

**Comment by Makman2** [comment]
oh 5* LOC^^

**Comment by sils1297** [comment]
If it's as easy as that, reopen, do and close it so we can all sleep well :)

**Comment by sils1297** [comment]
/me has absolutely no idea about what's happening here, he just randomly cites sentences from his in-memory-maintainer-manual :P

**Comment by Uran198** [comment]
I looked a little more in detail on this. This few lines will give you only
basic attributes completion. To complete bear names there will be a little
more work. I have wrote something, but it's breaking tests and I didn't
test it manually, for some reason. Also, there should be, probably,
documentation on how to set things up, so it works fine.

20:48, пт, 29.04.2016 Mischa Krüger <notifications@github.com>:
